### PR TITLE
Fix for #37, #50 and #53 (Windows specific issues)

### PR DESCRIPTION
**Pull request by engiecat**
While trying to use this repos on Windows, there were several issues, which were fixed in this PR.

1. #37 text_sequences are now converted into LongTensor, as required in index_select.
2. #50 Forced Garbage Collection can be done, and its probability can be set in hparams.py. This helps when num_workers=1 (if this is 2, MemoryError still occurs after approx. 2.5 hrs after training)
3. #53 when _frontend is None (as expected in Python Threading), getattr() injects _frontend to the thread.

This was tested on Windows 10 PC w/ GTX 1060 6GB, and the result of training is here.
(Several glitchy utterances are observed, but overall it is okay(is it??))

step000832500_text5_multispeaker10_alignment
https://www.dropbox.com/s/w8ynvpp7kcoxtyc/step000832500_text5_multispeaker10_predicted.wav?dl=0

step000832500_text3_multispeaker1_alignment
https://www.dropbox.com/s/12wtb9bzkl3bpdt/step000832500_text3_multispeaker1_predicted.wav?dl=0

step000832500_text0_multispeaker0_alignment
https://www.dropbox.com/s/yharzuw06orknbn/step000832500_text0_multispeaker0_predicted.wav?dl=0

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -129,7 +129,16 @@ def collect_features(self, *args):
             text, speaker_id = args
         else:
             text = args[0]
+        global _frontend # _frontend None bug fix
+        if _frontend == None:
+            _frontend = getattr(frontend, hparams.frontend)
         seq = _frontend.text_to_sequence(text, p=hparams.replace_pronunciation_prob)
+        _frontend = None # memory leaking prevention
```
Comment (r9y9) [inline train.py]: I think this is too tricky and I don't  want to do this on LInux, I'm ok on windows though. Could you add a branch to make sure this only happens on windows? e.g.,

```py
if platform.system() == "Windows":
    ....
```

File: train.py [hunk 2]
```diff
@@ -129,7 +129,16 @@ def collect_features(self, *args):
             text, speaker_id = args
         else:
             text = args[0]
+        global _frontend # _frontend None bug fix
+        if _frontend == None:
```
Comment (r9y9) [inline train.py]:
```py
if _frontend is None:
```

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,51 @@
+absl-py==0.1.10
```
Comment (r9y9) [inline requirements.txt]: The file should be deleted?

File: hparams.py [hunk 4]
```diff
@@ -251,10 +251,15 @@
     # 0 tends to prevent word repretetion, but sometime causes skip words
     window_backward=1,
     power=1.4,  # Power to raise magnitudes to prior to phase retrieval
+    
+    # GC:
+    # Forced garbage collection probability 
+    # Use only when MemoryError occurs
+    gc_probability=0.0002,
```
Comment (r9y9) [inline hparams.py]: I feel your fix is too tricky. I'd prefer to disable this by default.

**Comment by engiecat** [comment]
@r9y9 thank you for your reviews! I will test the proposed modifications and will update the PR!

**Comment by engiecat** [comment]
@r9y9  Testing done (this time for LJSpeech)!
It works well for now. 

step000210000_text0_single_alignment
https://www.dropbox.com/s/ww4d3g0xfq61svm/step000210000_text0_single_predicted.wav?dl=0

File: hparams.py [hunk 5]
```diff
@@ -251,8 +251,19 @@
...
+    # Use only when MemoryError continues in Windows (Disabled by default)
+    #gc_probability = 0.001,
 )
 
+# Preventing Windows specific error such as MemoryError 
+# Also reduces the occurrence of THAllocator.c 0x05 error in Widows build of PyTorch
+if platform.system() == "Windows":
+    print("Windows Detected - num_workers set to 1")
+    hparams.set_hparam('num_workers',1)
+
```
Comment (r9y9) [inline hparams.py]: Can you move this to `train.py`? Maybe next line of https://github.com/r9y9/deepvoice3_pytorch/blob/52d1026d477c39f2dd12761c461863e732ce86fc/train.py#L878?

Comment (engiecat) [reply hparams.py]: Okay I will :)

**Comment by mdda** [comment]
Why is requirements.txt being removed?

**Comment by r9y9** [comment]
Requirements are written in setup.py.

**Comment by engiecat** [comment]
@r9y9 There seems to be a conflict :/

**Comment by engiecat** [comment]
@r9y9 all done! was fumbling with github functionalities :(

File: train.py [hunk 6]
```diff
@@ -876,6 +888,16 @@ def restore_parts(path, model):
...
     hparams.parse(args["--hparams"])
+    
+    # Preventing Windows specific error such as MemoryError 
+    # Also reduces the occurrence of THAllocator.c 0x05 error in Widows build of PyTorch
+    if platform.system() == "Windows":
+        print("Windows Detected - num_workers set to 1")
+        hparams.set_hparam('num_workers',1)
+
+    # Now, print the finalized hparams.
+    print(hparams_debug_string())
```
Comment (r9y9) [inline train.py]: A really minor point though, `print(hparams_debug_string())` is called twice. I will fix it after merging.

**Comment by r9y9** [comment]
Thanks again!

**Pull request merged**